Fix some congestion control bugs and improve cwnd qlogging

agrover opened the pull request:
In diagnosing QNS Neqo server DC failures, I found a few bugs to fix.

Please see individual commit logs for details.

martinthomson commented on neqo-transport/src/cc.rs:
```diff
@@ -103,10 +99,16 @@ impl CongestionControl {
                 continue;
             }
+
+            if self.in_recovery {
+                self.in_recovery = false;
+                qlog::metrics_updated(&mut self.qlog, &[QlogMetric::InRecovery(false)]);
+            }
```
Can't we just set the value of congestion_recovery_start_time to `None` at this point?

ddragana replied:
I think we need to keep the value to be able to detect that a loss is in the old recovery phase even in case of reordering.

agrover replied:
Yeah this was the bug. We lost a packet, and therefore set CRST to now (T=1). Declaring another packet lost that was sent before CRST should not affect CC even if we have exited recovery (gotten an ack for a packet sent after CRST.)  Receiving an ack at T=2 for a packet sent after T=1 will cause CC to exit recovery. It's still possible for that or another ack received at T=3 to cause a sent packet(A), which previously didn't meet the packet reorder threshold (3), to be declared lost. If we forget previous CRST then declaring that packet lost will set CRST to T=3 and start a new recovery period, halves cwnd, etc. If CRST still is T=1 then we can tell packet A was sent before T=1.

The fact that CRST is an `Option<Instant>` is really only because there's no valid `Instant` value we can put there before the first time we lose a packet, causing recovery to happen.

martinthomson commented on neqo-transport/src/cc.rs:
```diff
@@ -220,27 +227,17 @@ impl CongestionControl {
     #[must_use]
-    fn in_congestion_recovery(&mut self, sent_time: Instant, packet_state: PacketState) -> bool {
+    fn in_congestion_recovery(&mut self, sent_time: Instant) -> bool {
         match self.congestion_recovery_start_time {
```
self.congestion_recovery_start_time.map_or(false, |t| sent_time <= t)

agrover replied:
Going to pass, I don't think this improves comprehension.

martinthomson commented on neqo-transport/src/recovery.rs:
```diff
@@ -729,6 +728,11 @@ impl LossRecovery {
+        // This must happen after on_packets_lost. If in recovery, this could
+        // take us out, and then lost packets will start a new recovery period
+        // when it shouldn't.
+        self.cc.on_packets_acked(&acked_packets);
```
👍 
This matches the spec.

ddragana commented:
A test is missing otherwise it is good.

agrover commented:
Rebased.

* Dropped commits merged as part of #881 
* Added commit with test
* Added commit with a renaming that I think helps understandability, but if you don't like it we can drop it.

martinthomson commented on neqo-transport/src/cc.rs:
```diff
@@ -302,3 +302,88 @@ impl CongestionControl {
+        let time_after1 = time_now + Duration::from_millis(100);
+        let time_after2 = time_now + Duration::from_millis(150);
+        let time_after3 = time_now + Duration::from_millis(175);
+        let pto = Duration::from_millis(100);
+        let rtt = Duration::from_millis(98);
```
const?
